java/core/src/java/org/apache/orc/impl/SchemaEvolution.java

```diff
@@ -45,16 +53,65 @@ public SchemaEvolution(TypeDescription readerSchema, boolean[] included) {
   }
 
   public SchemaEvolution(TypeDescription fileSchema,
-                         TypeDescription readerSchema,
-                         boolean[] included) throws IOException {
+      TypeDescription readerSchema,
+      boolean allowMissingMetadata,
+      boolean[] included) {
     readerToFile = new HashMap<>(readerSchema.getMaximumId() + 1);
     this.included = included;
     if (checkAcidSchema(fileSchema)) {
       this.readerSchema = createEventSchema(readerSchema);
     } else {
       this.readerSchema = readerSchema;
     }
-    buildMapping(fileSchema, this.readerSchema);
+
+    boolean useFieldNames;
+    if (!hasColumnNames(fileSchema)){
+      if (!allowMissingMetadata && !fileSchema.equals(readerSchema)){
+        throw new RuntimeException("Found that schema metadata is missing from file. This is likely caused by a writer earlier than HIVE-4243. Will not try to reconcile schemas");
+      } else {
+        LOG.warn("Found that schema metadata is missing from file. This is likely caused by a writer earlier than HIVE-4243. Will try to reconcile schemas based on index");
```

Comment:
How about "Found that column names are missing from file, which is caused by ORC writers before HIVE-4243."